File: codejam/2008/round_1a/minimum_scalar_product/solution.hs
import Data.List (intersperse, sort)

infixr 0 <|
(<|) f = f
(<||) f g x = f <| g x

solve :: IO String
solve = do
    xs <- fmap (map (read :: String -> Integer) <|| words) getLine
    ys <- fmap (map (read :: String -> Integer) <|| words) getLine
    return
        <| show
        <| sum
        <| map (uncurry (*)) <| zip (sort xs) (reverse <| sort ys)

main :: IO ()
main = (readLn :: IO Int)
    >>= sequence <|| flip replicate (getLine >> solve)
    >>= putStrLn
        <|| concat
        <|| intersperse "\n"
        <|| map (uncurry (++))
        <|| zip (map (("Case #" ++) <|| (++ ": ") <|| show) [1..])

File: codejam/2008/round_1a/minimum_scalar_product/solution.rs
lib::run!();

struct TestCase {
    v: Vec<i64>,
    w: Vec<i64>,
}

fn read() -> TestCase {
    lib::input!(_n);
    lib::input!(Vec<i64> as v);
    lib::input!(Vec<i64> as w);
    TestCase { v, w }
}

fn solve(TestCase { mut v, mut w }: TestCase) -> String {
    v.sort();
    w.sort();
    w.reverse();
    format!(
        "{}",
        v.into_iter()
            .zip(w.into_iter())
            .map(|(x, y)| x * y)
            .sum::<i64>()
    )
}
